### PR TITLE
docs: document the request-response cycle

### PR DESCRIPTION
**Pull request by hacksparrow**
Document the LB4 request-response cycle.
Closes https://github.com/strongloop/loopback-next/issues/4836

Signed-off-by: Yaapa Hage <hage.yaapa@in.ibm.com>

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by raymondfeng** [comment]
@hacksparrow It's a great start.

I would like to see more information, such as:

1. Which tier of the request/response trip is extensible/configurable?

2. How to add extensions to various tiers for different use cases? For example:
- authentication is enforced by an action in the sequence
- authorization is performed inside interceptors
- request body parsers
- ajv validation
- what about caching, rate-limiting, metrics, logging?
- error handling

3. More zoom-ins to various tiers to provide more details.

**Comment by raymondfeng** [comment]
This diagram can be useful too - https://loopback.io/pages/en/lb4/imgs/loopback-overview.png

**Comment by raymondfeng** [comment]
A few other things:

- How to access request/response objects
- How to propagate contextual information via DI

File: docs/site/Req-res-cycle.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+#### Request to a non-controller endpoint
+
+Non-controller endpoints can be created by static files, redirection rules,
+custom Express routes, and non-controller endpoints added by components.
+
+Requests to non-controller endpoints are also intercepted by the app's sequence,
+but the control is handed over to the underlying Express middleware from the
+`InvokeMethod` phase, and never make it to `Send` unless the middleware calls
+`next()`. Unhandled error from Express middleware still reach `Reject`.
```
Comment (agnes512) [inline docs/site/Req-res-cycle.md]: I like how we show the differences between these two endpoints. But I am a bit confused about `services` and the `interceptor` part. From my understanding, the structure we show above is:

```
             / ==> controller endpoint -> Services, Interceptors -> repositories
Sequence ---
             \ ==> non-controller endpoint
```

But IIUC I think the interceptors can also be use with non-controller endpoints. Maybe we make it more clear here.

Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Can you point out where this structure is contradicted?

```
             / ==> controller endpoint -> Services, Interceptors -> repositories
Sequence ---
             \ ==> non-controller endpoint
```

File: docs/site/Req-res-cycle.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+This method maps to the underlying Express' [static](https://expressjs.com/en/4x/api.html#express.static)
+method. It creates dynamic endpoints for files in the `rootDir` to be served
+at `path`.
+
+More details at [RestApplication.static()](https://loopback.io/doc/en/lb4/apidocs.rest.restapplication.static.html)
+doc.
+
+##### 2. app.redirect(fromPath, toPathOrUrl, statusCode)
+
+This method redirects requests to `fromPath` to `toPathOrUrl` with the status
```
[the remark this answers is not in the document]
Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Those parameter values, as shown in the signature on line 50.

File: docs/site/Req-res-cycle.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+Controller methods decorated with operation decorators like `@get()`, `@post()`,
+`@put` etc., create endpoints on the app for the corresponding HTTP verbs. The
+behavior of these endpoints are entirely dependent on the implementation of the
+controller method.
+
+{% include tip.html content="Apart from controller files in the `controllers` directory,
+controllers may be added to the app by [components](https://loopback.io/doc/en/lb4/Components.html)." %}
+
+In the request-response cycle section we will see how implemenation details
+determine the course of a request these endpoints - they may or may not actually
```
[the remark this answers is not in the document]
Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Errata, should be: determine the course of a request to these endpoints

File: docs/site/Req-res-cycle.md (hunk 1 above)
Comment (agnes512) [reply docs/site/Req-res-cycle.md]: It looks like the interceptors cannot be used in non-controller endpoints. Is it true?

Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Interceptors work only on Controllers and Repositories, so non-controller endpoint handlers can't be intercepted.

Comment (agnes512) [reply docs/site/Req-res-cycle.md]: image
I think I misunderstood the hander in the diagram as the non-controller endpoints.

Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Works on Services too.

**Comment by dhmlau** [comment]
@hacksparrow , would this PR `closes #1833` instead of `Closes #4836`, if this PR is only about LB4 req/res cycle?

File: docs/site/Req-res-cycle.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+[Components](https://loopback.io/doc/en/lb4/Components.html) are the extension
+units of LoopBack. Since they can access the LoopBack app instance, the request
+and response instances using [dependency injection](https://loopback.io/doc/en/lb4/Dependency-injection.html),
+they are capable of adding new endpoints and determining the result of a request.
+
+### The request-response cycle
+
+The request-response cycle involves many components, especially if the request
+is to a controller-endpoint which interacts with a model.
```
Comment (dhmlau) [inline docs/site/Req-res-cycle.md]: `the request is to a controller-endpoint` -> `the request is coming from a controller-endpoint`?

Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Made it "interacts with the database".

**Comment by hacksparrow** [comment]
@dhmlau only LB4 req-res.

**Comment by hacksparrow** [comment]
@raymondfeng I have updated the PR with your inputs. However, since this PR is only for the flow of the request and the response, I have only mentioned your points from a high-level perspective.

Your points should be covered in another doc, maybe titled "Writing extensions for the request-response cycle".

File: docs/site/Req-res-cycle.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
+---
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: Nitpick: let's name the file as `Request-response-cycle.md`.

File: docs/site/Req-res-cycle.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+lang: en
+title: 'Request-response cycle'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Application.html
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: Wrong html link. It should be `/doc/en/lb4/Request-response-cycle.html`.

File: docs/site/Req-res-cycle.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+---
+
+## The LoopBack 4 request-response cycle
+
+To understand the LoopBack 4 request-response cycle, let's start by enumerating
+the APIs that create the endpoints on the server. We will then follow the path
+taken by a request, to see how it makes its way through the various parts of the
+framework to return a result.
+
+### Setting up the request-responses infrastruture
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: `request-responses` -> `request/response`

File: docs/site/Req-res-cycle.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
+---
+lang: en
+title: 'Request-response cycle'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Application.html
+---
+
+## The LoopBack 4 request-response cycle
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: Maybe `request/response`?

File: docs/site/Req-res-cycle.md [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+it is the gatekeeper of all requests to the app. Every request, whether to
+controller endpoints or non-controller endpoints, must pass through the sequence.
+
+The sequence identifies the responsible handler for the requested endpoint and
+passes on the request to the handler. The handler then take care of sending
+the response back to the client.
+
+#### The sequence
+
+The sequence is a simple class with five injected helper methods. These five
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: `with five injected helper methods` -> `with five injected helper methods by default`?

File: docs/site/Req-res-cycle.md [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+The `PingController` controller has a method named `ping()`, decorated with a
+`get()` decorator. This setup creates the `/ping` endpoint on the app.
+
+The app is configured to use the `MySequence` sequence.
+
+Now this is a step-wise sequence of what happens when a request is made to
+`http://localhost:4000/ping`.
+
+1. The request is intercepted by `MySequence`.
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: `is intercepted by` -> 'is dispatched to MySequence by the http server`?

File: docs/site/Req-res-cycle.md [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+2. The route is then identified by `FindRoute`. There is a handler for this request.
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: based on the verb and path defined by OpenAPI decorators?

File: docs/site/Req-res-cycle.md [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+3. `ParseParams` then tries to parses any parameters might have been submitted.
+There is none.
+4. `InvokeMethod` then invokes the handler with empty parameters. It returns
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: `empty parameters` -> `resolved parameters which is empty in this case`?

File: docs/site/Req-res-cycle.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+whatever the handler returns, in this case, it is JSON object with `greeting`,
+`date`, `url`, and `headers` properties.
+5. `Send` then sends this JSON object back to the client.
```
Comment (raymondfeng) [inline docs/site/Req-res-cycle.md]: Let's mention error handling here for `Reject`.

File: docs/site/Req-res-cycle.md (hunk 1 above)
Comment (raymondfeng) [reply docs/site/Req-res-cycle.md]: IIRC, global interceptors also intercept non-controller routes.

File: docs/site/Req-res-cycle.md (hunk 9 above)
Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Made it `five injected helper methods in the constructor.`

File: docs/site/Req-res-cycle.md (hunk 13 above)
Comment (hacksparrow) [reply docs/site/Req-res-cycle.md]: Added: `6. `Reject` is in place to catch any errors that might be thrown in any
of the steps above.`

**Comment by hacksparrow** [comment]
@raymondfeng applied all your feedback. PTAL.

**Comment by raymondfeng** [comment]
@hacksparrow Please fix the [formatting error](https://travis-ci.com/github/strongloop/loopback-next/jobs/302708425) with `npm run prettier:fix`.

**Comment by dhmlau** [comment]
Merging the PR per @hacksparrow's request.

**Pull request merged**